text/0000-afp.md

```diff
@@ -0,0 +1,354 @@
+  throw new Error('Extended classes must implement onHireConfirmed')
+}
+
+/**
+ * On receipt of a reward receipt from a farmer.
+ * @param {messages.Receipt} receipt
+ * @param {PeerConnection} connection
+ */
+async onReceipt(receipt, connection) {
+  throw new Error('Extended classes must implement onReceipt')
+}
+```
+
+#### 4.2.2. Farmer
+
+A farmer must extend the Farmer class to define that farmer’s specifications for generating a quote for a task, validating a SOW for a task, and signing and validating an agreement for a task. If rewards are expected, then a farmer would also handle validating the reward and generating a receipt.
+
+```js
+/**
+ * Return whether a sow is valid.
+ * @param {messages.SOW} sow
+ * @returns {boolean}
+ */
+async validateSow(sow) {
+  throw new Error('Extended classes must implement validateSow.')
+}
+
+/**
+ * Return whether a reward is valid.
+ * @param {messages.Reward} reward
+ * @returns {boolean}
+ */
+async validateReward(reward) {
+  throw new Error('Extended classes must implement validateReward.')
+}
+
+/**
+ * Return whether an agreement is valid.
+ * @param {messages.Agreement} agreement
+ * @returns {boolean}
+ */
+async validateAgreement(agreement) {
+  throw new Error('Extended classes must implement validateAgreement.')
+}
+
+/**
+ * Return a receipt given a reward.
+ * @param {messages.Reward} reward
+ * @returns {messages.Receipt}
+ */
+async generateReceipt(reward) {
+  throw new Error('Extended classes must implement generateQuote.')
+}
+
+/**
+ * Return a quote given a sow.
+ * @param {messages.SOW} sow
+ * @returns {messages.Quote}
+ */
+async generateQuote(sow) {
+  throw new Error('Extended classes must implement generateQuote.')
+}
+
+/**
+ * Sign and return an agreement.
+ * @param {messages.Agreement} agreement
+ * @returns {messages.Agreement}
+ */
+async signAgreement(agreement) {
+  throw new Error('Extended classes must implement signAgreement.')
+}
+```
+
+#### 4.2.3. Matcher
+
+Different service requesters may have different needs when selecting peers, such as selecting the cheapest set, the fastest set, the first set number of peers, etc. To allow for this, each service may implement their own matcher (or use one of a set of predefined matchers) that extends the Matcher class. This class describes an object that, given a set of options, selects a subset of peers using a matching strategy specific to the service.
+
+```js
+/**
+ * Add a quote for consideration. If a quote is considered
+ * valid, then call hireFarmerCallback to continue
+ * agreement process.
+ * @param {messages.Quote} quote
+ * @param {function(messages.Agreement)} hireFarmerCallback
+ */
+async addQuote(quote, hireFarmerCallback) {
+  throw new Error('Extended classes must implement addQuote')
+}
+
+/**
+ * Remove quote from consideration.
+ * @param {messages.Quote} quote
+ */
+async removeQuote(quote) {
+  throw new Error('Extended classes must implement removeQuote')
+}
+```
+
+## 5. Real World Example
```

Comment:
I know it might be a lot to put code here implementing all the above classes/functions, but maybe a link to a gist of them?  Or put it here, just thinking of the example in 0004-afs-contract-standard.